### DISCUSSION TITLE
Switching language

### DISCUSSION THREAD

**Discussion by dengxk1234** [end-4/dots-hyprland#372] [Q&A]
I'm Chinese and used to using interfaces in Chinese. I've tried to change `/etc/locale.conf` but it only changes the time format, so is there any way to switch the language of hyprland, ags, etc. to Chinese?

**Comment by end-4** [comment]
for typing you can use fcitx5
as for the interface language, there isn't a way to do that (yet)

**Comment by clsty** [comment]
About Hyprland: It merely output text on the UI (unless for giving warning, or in the terminal using Hyprctl etc.)
So it's not very useful to have multi-language support for it.

---

About AGS: Multi-language support could be considered, but it needs implemention by devs (@end-4 or @Aylur).

Generally it contains two parts: 
- i18n, i.e. providing a standard method to store translation and showing translated text according to locale (on gtk/AGS/dots-hyprland side, need help from the devs)
- l10n, i.e. translation (need help from contributors using variant languages)

Additional info on i18n:
I recommend using "po" file (editable with "poedit") as it's a common thing, also used by `archinstall`, `paru`, etc.
And it's easy to integrated with translation platforms, e.g. weblate, so that contributors will be easier to translate the text with a friendly and powerful interface on the web.

---

Others:
We already have a Wiki site that supports multiple languages. It's l10n part is open to contribution, see [here](https://end-4.github.io/dots-hyprland-wiki/en/dev/doc-site-contrib/) if you're interested.

**Reply by end-4** [reply]
Language support has nothing to do with Aylur. His project is a widget library, not an UI translator

**Reply by end-4** [reply]
I've though of language support but
- Idk it well and don't know if it'll be a lot of work to get it to work properly
- Don't know if it'll be *that* helpful. I mean, "China has many people" is a valid argument but it's just very naive lol

**Reply by clsty** [reply]
Welp, AGS does shows some text on the interface, isn't it?
As far as I know, the action "showing" is often related to i18n support, if it has one.
E.g. for bash script, `gettext` is often used for i18n support and used to show text (it does not support color well so I'm still experimenting on it. i.e. i18n is not easy for installation script currently).
Not sure though. Correct me if I'm wrong.

**Reply by end-4** [reply]
the solution that i would be absolutely happy to implement is a js file containing ui strings - that's super simple, but won't be very standard(?)

**Reply by clsty** [reply]
I agree with you that it may not be very helpful. The total quantity of the text on the UI is not very large.

However, i18n is not only for Chinese, but for any non-English users and will sure make this project more friendly to them.
(I'm relatively used to English UI so I don't mind it for myself.)

Anyway, it's an optional thing to be implemented or not.

**Reply by clsty** [reply]
"but won't be very standard(?)"
Yeah, that's exactly what I meant.
i18n should provide a "standard" way. That's why I think it maybe could be supported on AGS part.

**Comment by clsty** [comment]
And an offtopic question (just out of curiosity):

I've noticed that the stars of this repo increased about 300 in several days.
So, how do you get to know this project?
I mean, maybe someone has been using this repo and recommended it in some chat group, etc.?

**Reply by end-4** [reply]
@yrnmsk told me about this thread which contains a link to this repo: https://news.ycombinator.com/item?id=39803287
though i don't use Hacker News and don't know how popular it is

The "Insights" tab of the repo doesn't report visits from that site

**Reply by end-4** [reply]
btw
if you're curious about the huge increase earlier, it was thanks to this youtube video 
![image](https://github.com/end-4/dots-hyprland/assets/97237370/514c1548-7519-4a99-837d-8b3ea569eb8f)

**Reply by ghost** [reply]
Yeah.. I'm not sure if that site is popular as well or not, but iirc, I did see someone say that that article was apparently on the front page at that time, so I guess there's that?

Edit:
Found the Luyu's comment that initially shared the hn thread

![image](https://github.com/end-4/dots-hyprland/assets/72034970/77954b59-47c5-4e73-82f1-3a600748a2be)

But I think the youtube one @end-4 just sent prolly did more work than the hn thread overtime. There's actually 2 videos from the same youtube channel that have at least ~2k viewers featuring end's rice.

**Comment by dengxk1234** [comment]
Thanks for the quick response. I just want to know if there is a way to change the language by myself. Since you tell me that it need the developer to work with it and this thought is naive, there's no need to trouble them and I'll just use the English interface.
PS:I'm looking for a good configuration for hyprland so I found this project from an article on the web.

**Reply by clsty** [reply]
Can you please give me the link to the article?

**Comment by dengxk1234** [comment]
One last question: I'm in China and the network can't reach Google, OpenAI, etc. so I have to use a proxy server only for the AI sidebar. Is there a way to configure it simply?

**Reply by end-4** [reply]
Unfortunately not
But feel free to suggest a service that's available in China without the need for such setups, I'll be happy to add it

**Reply by end-4** [reply]
Bing? I see you have a fork of an unofficial bing chat api library...

**Reply by dengxk1234** [reply]
Oh, that's just for test and deploying usage and it can't work properly in China.

**Reply by dengxk1234** [reply]
I've found some ai providers and don't know if they can work.
1. There are some ChatGPT reversed projects or those with free keys on github.
2. There are some mirror sites which can work in China.
3. THUDM has trained a model called ChatGLM which can work in China provided both offline and online([offline model](https://github.com/THUDM/ChatGLM-6B) or [online api](https://open.bigmodel.cn/dev/api#glm-3-turbo)).

**Reply by end-4** [reply]
# Option 1 and 2
- That's vague, i don't know those "reversed" stuff and "mirrors" `¯\_(ツ)_/¯` 
# Option 3
## Offline
Unfortunately my machine isn't powerful enough to run locally
## Online
It isn't free, so i can't test it. But in case you're able to access it, you can try adding this to the `PROVIDERS` object in `~/.config/ags/services/gpt.js`
```js
    'glm_zhipu': {
        'name': 'ChatGLM - Zhipu AI',
        'logo_name': 'ai-oxygen-symbolic',
        'description': '',
        'base_url': 'https://open.bigmodel.cn/api/paas/v4/chat/completions',
        'key_get_url': '',
        'key_file': 'zhipu_key.txt',
    },
```
then restart ags (`pkill ags; ags`), go to the GPTs tab of sidebar and select the new entry you added in the providers dropdown menu 
> [!NOTE] 
> You can do this for other APIs, if you happen to find any. Just change the `base_url` (the endpoint) and `key_file` (the name of the file in which you store your key, located in `~/.cache/ags/user/ai`)

**Comment by dengxk1234** [comment]
The problem of changing the proxy server is solved in [this PR](https://github.xyz/end-4/dots-hyprland/pull/396). Thanks to the developers.
